File: publicChat_web/pages/types/message.ts
type Message = {
    "SendUser": string, // 发送人
    "SendTime": string, // 发送时间
    "Msg": string, //  消息内容
    "IsSend": boolean, //  // 是否是发送消息，用于区分发送消息和上线下线消息（true 发送消息 false 上线/下线消息）
    "IsPublic": boolean, // 消息类型是否是公开的 true 公开 false 私信
    "ReceUser": string, // 接收人
    "IsImg": boolean, // 头像是否是图片
    "Pic": string, // 头像图片地址
}

File: publicChat_backend/message.go
package main

import (
	"log"

	"github.com/gorilla/websocket"
)

// 公共消息
func publicMessage(msg Msg) {
	for _, user := range users {
		// 当 msg.IsSend 为true时，说明是发送消息，则必须判断 user.Name != msg.SendUser
		if user.Conn != nil && ((msg.IsSend && user.Name != msg.SendUser) || !msg.IsSend) {
			err := user.Conn.WriteMessage(websocket.TextMessage, msg.EncodeMessage())
			if err != nil {
				log.Println(err)
			}
		}
	}
}

// 发送私聊消息给指定用户
func privateMessage(msg Msg) {
	for _, user := range users {
		if user.Name == msg.ReceUser && user.Conn != nil { // 当接收人在线时
			// 发送私聊消息
			err := user.Conn.WriteMessage(websocket.TextMessage, msg.EncodeMessage())
			if err != nil {
				log.Println(err)
			}
			msg.IsRece = true // 将 IsRece 设置为true
			break
		}
	}
	if !msg.IsRece { // 只有接收人离线时，才将消息存到消息列表中
		msgs.PushBack(msg)
	}
}
